etc/emacstool.1
.TH EMACSTOOL 1
.SH NAME
.I emacstool 
\- run emacs under Sun windows with function-key and mouse support.
.SH SYNOPSIS
.I emacstool
[{window_args} {-rc run_command_path} args ... ]
.SH TYPICAL USAGE
In ~/.suntools or ~/.rootmenu include a line like this:
.br
"Emacstool"     emacstool -WI emacs.icon -f emacstool-init
.SH DESCRIPTION
.B  Emacstool
creates a SunView frame and a tty subwindow within which mouse events
and function keys are translated to ASCII sequences which Emacs can
parse.  The translated input events are sent to the process running in
the tty subwindow, which is typically GNU Emacs.  Emacstool thereby
allows GNU Emacs users to make full use of the mouse and function keys.
GNU Emacs can be loaded with functions to interpret the mouse and
function-key events to make a truly fine screen oriented editor for
the Sun Workstation.
.PP
(Note that GNU Emacs has a special interface to the X window system as
well.  The X window system has many technical advantages, it is an
industry standard, and it is also free software.  The Free Software
Foundation urges you to try X windows, and distributes a free copy of
X on Emacs distribution tapes.)
.PP
Function keys are translated to a sequence of the form
`^X*[a-o][lrt]'.  The last character is `l', `r', or `t' corresponding
to whether the key is among the Left, Right, or Top function keys.
The third character indicates which button of the group
was pressed.  Thus, the function key in the lower right corner will
transmit the sequence `^X*or'.  In addition, the [lrt] is affected by
the Control, Meta, and Shift keys.  Unshifted Control keys will be
non-alphabetic: C-l is [,], C-r is [2], C-t is [4].
.PP 
Mouse buttons are encoded as `^X^@([124] x y)\\n'.  ^X^@ is the
standard GNU Emacs mouse event prefix, it is followed by a list
indicating the button pressed and the character row and column of the
point in the window where the mouse cursor is, and followed by a
newline character.  In GNU Emacs, the ^X^@ dispatches to a
mouse event handler which then reads the following list.
.SH OPTIONS
.B Emacstool
supports all the standard window arguments, including font and icon 
specifiers.  
.PP
By default, Emacstool runs the program 
.I emacs
in the created subwindow.  
The value of the environment variable 
.I EMACSTOOL
can be used to override this if your version of 
.B Emacs
is not accessible on your search path by the name 
.I Emacs.
In addition, the run command can be set by the 
.I pathname 
following the last occurrence of the
.I \-rc
flag.
This is convenient for using Emacstool to run on remote machines.
.PP
All other command line arguments not used by the window system are passed
as arguments to the program that runs in the Emacstool window.
.PP
For example: 
.PP
local% (emacstool -rc rlogin remote -8 &)&
.PP
will create an Emacstool window logged in to a machine named
.I remote.
If Emacs is run from this window, 
Emacstool will encode mouse and function keys, and send them to rlogin.
If Emacs is run from this shell on the remote machine, it will see
the mouse and function keys properly.
However, since the remote host does not have access to the screen,
the cursor cannot be changed, menus will not appear, and the selection
buffer (STUFF) is limited.
.SH Using With GNU Emacs:
The GNU Emacs files
lisp/term/sun.el,
lisp/sun-mouse.el, 
lisp/sun-fns.el, 
and 
src/sunfns.c
provide emacs support for the Emacstool and function keys.
Emacstool will automatically set the TERM environment variable to be "sun"
and unset the environment variable TERMCAP.  That is, these variables will
not be inherited from the shell that starts Emacstool.
Since the terminal type is
.I SUN
(that is, the environment variable TERM is set to 
.I SUN), 
Emacs will automatically load the file lisp/term/sun.
This, in turn, will ensure that sun-mouse.el is autoloaded when any mouse
events are detected.  It is suggested that 
.I sun-mouse 
and
.I sun-fns
be loaded in your site-init.el file, so that they will always be loaded
when running on a Sun workstation.
.PP
In addition, Emacstool sets the environment variable IN_EMACSTOOL = "t".
Lisp code in your ~/.emacs can use (getenv "IN_EMACSTOOL")
to determine whether to do Emacstool specific initialization.
Sun.el uses this to automatically call emacstool-init (getenv "IN_EMACSTOOL")
is defined.
.PP
The file src/sunfns.c defines several useful functions for emacs on
the Sun.  Among these are procedures to pop-up SunView 
.I menus, 
put and get from the SunView
.I STUFF
buffer, and a procedure for changing the cursor 
.I icon.
If you want to define or edit cursor icons, 
there is a rudimentary mouse driven icon editor in the file
lisp/sun-cursors.el.  Try invoking (sc:edit-cursor)
.SH BUGS
It takes a few milliseconds to create a menu before it pops up.
.SH ENVIRONMENT VARIABLES
EMACSTOOL
IN_EMACSTOOL
TERM
TERMCAP
.SH FILES
.DT
emacs
.SH "SEE ALSO"
emacs(1)
 .../etc/SUN-SUPPORT
 .../lisp/term/sun.el
.SH COPYING
Copyright 
.if t \(co
.if n (c)
2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008  Free Software Foundation, Inc.
.PP
Permission is granted to make and distribute verbatim copies of this
document provided the copyright notice and this permission notice are
preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of
this document under the conditions for verbatim copying, provided that
the entire resulting derived work is distributed under the terms of
a permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
document into another language, under the above conditions for
modified versions, except that this permission notice may be stated
in a translation approved by the Free Software Foundation.

.\" arch-tag: a1f44815-5163-4026-89c1-4404e4b81d37


etc/NEWS
GNU Emacs NEWS -- history of user-visible changes.

Copyright (C) 2007, 2008 Free Software Foundation, Inc.
See the end of the file for license conditions.

Please send Emacs bug reports to emacs-pretest-bug@gnu.org.
If possible, use M-x report-emacs-bug.

This file is about changes in Emacs version 23.

See files NEWS.22, NEWS.21, NEWS.20, NEWS.19, NEWS.18, and NEWS.1-17
for changes in older Emacs versions.

You can narrow news to a specific version by calling `view-emacs-news'
with a prefix argument or by typing C-u C-h C-n.

Temporary note:
 +++ indicates that the appropriate manual has already been updated.
 --- means no change in the manuals is called for.
When you add a new item, please add it without either +++ or ---
so we will look at it and add it to the manual.


* About external Lisp packages


* Installation Changes in Emacs 23.1

** The default X toolkit is now Gtk+, rather than Lucid.

** The Mac Carbon port is no longer supported.
Instead, use... [what?]

** configure now checks for libgif before libungif when searching for
a GIF library.

** Support for systems without alloca has been removed.

** Support for Sun windows has been removed.

** Support for many obsolete platforms has been removed.
See the list at the end of etc/MACHINES for details.

** The `emacstool' utility has been removed.

** The new configuration option "--with-dbus" enables D-Bus language
bindings for Emacs.

* Changes in Emacs 23.1

** Emacs now supports using both X displays and ttys in one session.
You can also use any number of different ttys.

You can test for the presence of this feature in your Lisp code by
testing for the `multi-tty' feature.

** Emacs comes with a new set of icons for Mac OS X.
OS-X-style icons (an application icon and a relevant document icon)
were contributed by Kentaro Ohkouchi.
Source files for these icons can be found in Emacs.app/Contents/Resources.
PNG versions are available as etc/images/icons/emacs*_mac.png.

** Built-in functions (subr) can now have an interactive specification
that is not a prompt string.  If the `intspec' parameter of a `DEFUN'
starts with a `(', the string is evaluated as a Lisp form.

** set-file-modes is now interactive and can take the mode value in
symbolic notation thanks to auxiliary functions.

** split-window-preferred-function specifies whether display-buffer should
split windows vertically or horizontally.

** Emacsclient has been extended to support opening a new terminal
frame.  Its behavior has been changed to open a new Emacs frame by
default.  Use the -c option to get the old behavior of opening files in
the currently selected Emacs frame.

** The refcards are now shipped as PDF files.

** Emacs now supports the SVG image format through librsvg2.

** If you set find-file-confirm-nonexistent-file to t, then C-x C-f
requires confirmation before opening a non-existent file.

** If the gpm mouse server is running and t-mouse-mode enabled, Emacs uses a
Unix socket in a GNU/Linux console to talk to server, rather than faking events
using the client program mev.  This C level approach provides mouse
highlighting, and help echoing in the minibuffer.

** The new variable next-error-recenter specifies how next-error should
recenter the visited source file.  Its value can be a number (for example,
0 for top line, -1 for bottom line), or nil for no recentering.

** The mode-line displays a `@' if the default-directory for the current buffer
is on a remote machine, or a hyphen otherwise.

** The mode-line displays a mode menu when mouse-1 is clicked on a minor mode,
in the same way as it already did for major modes.

** The new command balance-windows-area balances windows both vertically
and horizontally.

** The new command close-display-connection can be used to close a connection
to a remote display, e.g. because the display is about to become unreachable.

** The command shell prompts for the default directory, when it is
called with a prefix, and the default directory is a remote file name.
This is because some file name handlers (like ange-ftp) are not able to
run processes remotely.

** The new command `display-time-world' starts an updating time display
using several time zones, in a buffer.

* Startup Changes in Emacs 23.1

** New user option `initial-buffer-choice' specifies what to display
after starting Emacs: startup screen, *scratch* buffer, visiting a
file or directory.

** New alias `argv' for `command-line-args-left'
This is a convenience alias, so that one can write `(pop argv)'
inside of --eval command line arguments in order to access
following arguments.

** The abbrev file is no longer read at startup in batch mode.

* Incompatible Editing Changes in Emacs 23.1

+++
** In Dired-x, all command guesses for ! are now added to the default
list accessible by M-n instead of pushing all guesses temporarily into
the history list.


* Editing Changes in Emacs 23.1

+++
** C-M-% now shows replacement as it would look in the buffer, with
`\N' and `\&' substituted according to the match.  Old behavior can be
restored by customizing `query-replace-show-replacement'.

+++
** M-q now fills the region if the region is active and
`transient-mark-mode' is turned on.  Otherwise, it fills the current
paragraph.

+++
** M-$ now checks spelling of the region if the region is active and
`transient-mark-mode' is turned on.  Otherwise, it checks spelling of the
word at point.

** TAB now indents the region if the region is active and
`transient-mark-mode' is turned on.

** `use-empty-active-region' controls whether an empty active region
in Transient Mark mode should make commands operate on that empty region.

** C-z now invokes `suspend-frame', C-x C-c now invokes
`save-buffers-kill-terminal'.

** New command kill-matching-buffers kills buffers whose name matches a regexp.

** You can disable kill ring commands from accessing the primary selection
by setting `x-select-enable-primary' to nil.


lib-src/emacstool.c
/*
   Copyright (C) 1986, 1988, 1990, 1991, 2001, 2002, 2003, 2004,
                 2005, 2006, 2007, 2008  Free Software Foundation, Inc.

This file is part of GNU Emacs.

GNU Emacs is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3, or (at your option)
any later version.

GNU Emacs is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with GNU Emacs; see the file COPYING.  If not, write to
the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
Boston, MA 02110-1301, USA.  */

/*
 * For Emacs in SunView/Sun-Windows: (supported by Sun Unix v3.2 or greater)
 * Insert a notifier filter-function to convert all useful input
 * to "key" sequences that emacs can understand.  See: Emacstool(1).
 *
 * Author: Jeff Peck, Sun Microsystems, Inc. <peck@eng.sun.com>
 *
 * Original Idea: Ian Batten
 * Updated 15-Mar-88, Jeff Peck: set IN_EMACSTOOL, TERM, TERMCAP
 * Updated 10-Sep-88, Jeff Peck: add XVIEW and JLE support
 * Updated  8-Oct-90, Jeff Peck: add Meta-bit for Xview
 * Updated  6-Mar-91, Jeff Peck: Hack to detect -Wt invocation
 *	[note, TTYSW limitation means you must Click-To-Type in Openwin]
 *	[fixed in OW3 or use local/tty.o]
 *	for better results, this should move to using TERMSW.
 * Updated 10-Mar-91, Jeff Peck, et al: support for TERMSW (TTERM)
 *	allows point-to-type even in OW2
 *
 * 	[note: xvetool should be started with the "-nw" flag for emacs!]
 */

#ifdef XVIEW
#include <xview/xview.h>
#include <xview/panel.h>
#include <xview/attr.h>
#include <xview/tty.h>
#include <xview/ttysw.h>		/* private defines */
#include <xview/termsw.h>		/* -DTTERM */
#include <xview/font.h>			/* for testing */
#else
#include <suntool/sunview.h>
#include <suntool/tty.h>
#include <suntool/ttysw.h>
#endif XVIEW

#ifdef JLE
# include <locale.h>
#endif JLE

#include <stdio.h>
#include <sys/file.h>

#define BUFFER_SIZE 128               /* Size of all the buffers */

/* define WANT_CAPS_LOCK to make f-key T1 (aka F1) behave as CapsLock */
#define WANT_CAPS_LOCK
#ifdef WANT_CAPS_LOCK
int caps_lock;		/* toggle indicator for f-key T1 caps lock */
static char *Caps = "[CAPS] ";		/* Caps Lock prefix string */
#define CAPS_LEN 7			/* strlen (Caps) */
#endif

static char *mouse_prefix = "\030\000";	/* C-x C-@ */
static int   m_prefix_length = 2;       /* mouse_prefix length */

static char *key_prefix = "\030*";  	/* C-x *   */
static int   k_prefix_length = 2;       /* key_prefix length */

#ifdef JLE
static char *emacs_name = "nemacs";	/* default run command */
static char *title = "NEmacstool - ";	/* initial title */
#else
static char *emacs_name = "emacs";	/* default run command */
static char *title = "Emacstool - ";	/* initial title */
#endif JLE

static char buffer[BUFFER_SIZE];	/* send to ttysw_input */
static char *bold_name = 0;	 	/* for -bold option */

Frame frame;                            /* Base frame for system */

#ifndef TTERM
#define SWTYPE TTY
Tty tty_win;				/* Where emacs is reading */
#else
#define SWTYPE TERMSW
Termsw tty_win;				/* Termsw does follow-mouse */
#endif TTERM

#ifdef XVIEW
Xv_Window tty_view;			/* Where the events are in Xview*/
#else
Tty tty_view;				/* SunView place filler */
#endif XVIEW

int font_width, font_height;            /* For translating pixels to chars */
int left_margin = 0;		/* default window -- frame offset */

int console_fd = 0;		/* for debugging: setenv DEBUGEMACSTOOL */
FILE *console;			/* for debugging: setenv DEBUGEMACSTOOL */

Icon frame_icon;
/* make an icon_image for the default frame_icon */
static short default_image[258] =
{
#include <images/terminal.icon>
};
mpr_static(icon_image, 64, 64, 1, default_image);

/*
 * Assign a value to a set of keys
 */
int
button_value (event)
     Event *event;
{
  int retval = 0;
  /*
   * Code up the current situation:
   *
   * 1 = MS_LEFT;
   * 2 = MS_MIDDLE;
   * 4 = MS_RIGHT;
   * 8 = SHIFT;
   * 16 = CONTROL;
   * 32 = META;
   * 64 = DOUBLE;
   * 128 = UP;
   */

  if (MS_LEFT   == (event_id (event))) retval = 1;
  if (MS_MIDDLE == (event_id (event))) retval = 2;
  if (MS_RIGHT  == (event_id (event))) retval = 4;

  if (event_shift_is_down (event)) retval += 8;
  if (event_ctrl_is_down  (event)) retval += 16;
  if (event_meta_is_down  (event)) retval += 32;
  if (event_is_up         (event)) retval += 128;
  return retval;
}

/*
 *  Variables to store the time of the previous mouse event that was
 *  sent to emacs.
 *
 *  The theory is that to time double clicks while ignoring UP buttons,
 *  we must keep track of the accumulated time.
 *
 *  If someone writes a SUN-SET-INPUT-MASK for emacstool,
 *  That could be used to selectively disable UP events,
 *  and then this cruft wouldn't be necessary.
 */
static long prev_event_sec = 0;
static long prev_event_usec = 0;

/*
 *  Give the time difference in milliseconds, where one second
 *  is considered infinite.
 */
int
time_delta (now_sec, now_usec, prev_sec, prev_usec)
     long now_sec, now_usec, prev_sec, prev_usec;
{
  long sec_delta = now_sec - prev_sec;
  long usec_delta = now_usec - prev_usec;

  if (usec_delta < 0) {		/* "borrow" a second */
    usec_delta += 1000000;
    --sec_delta;
  }

  if (sec_delta >= 10)
    return (9999);		/* Infinity */
  else
    return ((sec_delta * 1000) + (usec_delta / 1000));
}


/*
 * Filter function to translate selected input events for emacs
 * Mouse button events become ^X^@(button x-col y-line time-delta) .
 * Function keys: ESC-*{c}{lrt} l,r,t for Left, Right, Top;
 * {c} encodes the keynumber as a character [a-o]
 */
static Notify_value
input_event_filter_function (window, event, arg, type)
#ifdef XVIEW
     Xv_Window window;
#else
     Window window;
#endif XVIEW
     Event *event;
     Notify_arg arg;
     Notify_event_type type;
{
  struct timeval time_stamp;

  if (console_fd) fprintf(console, "Event: %d\n", event_id(event));

  /* UP L1 is the STOP key */
  if (event_id(event) == WIN_STOP) {
    ttysw_input(tty_win, "\007\007\007\007\007\007\007", 7);
    return NOTIFY_IGNORED;
  }

  /* UP L5 & L7 is Expose & Open, let them pass to sunview */
  if (event_id(event) == KEY_LEFT(5) || event_id(event) == KEY_LEFT(7))
    if(event_is_up (event))
      return notify_next_event_func (window, event, arg, type);
    else return NOTIFY_IGNORED;

  if (event_is_button (event)) { 	      /* do Mouse Button events */
/* Commented out so that we send mouse up events too.
   if (event_is_up (event))
      return notify_next_event_func (window, event, arg, type);
*/
    time_stamp = event_time (event);
    ttysw_input (tty_win, mouse_prefix, m_prefix_length);
    sprintf (buffer, "(%d %d %d %d)\015",
	     button_value (event),
	     (event_x (event) - left_margin) / font_width,
	     event_y (event) / font_height,
	     time_delta (time_stamp.tv_sec, time_stamp.tv_usec,
			 prev_event_sec, prev_event_usec)
	     );
    ttysw_input (tty_win, buffer, strlen(buffer));
    prev_event_sec = time_stamp.tv_sec;
    prev_event_usec = time_stamp.tv_usec;
    return NOTIFY_IGNORED;
  }

  { /* Do the function key events */
    int d;
    char c = (char) 0;
    if ((event_is_key_left  (event)) ?
	((d = event_id(event) - KEY_LEFT(1)   + 'a'), c='l') :
	((event_is_key_right (event)) ?
	 ((d = event_id(event) - KEY_RIGHT(1) + 'a'), c='r') :
	 ((event_is_key_top   (event)) ?
	  ((d = event_id(event) - KEY_TOP(1)  + 'a'), c='t') : 0)))
      {
	if (event_is_up(event)) return NOTIFY_IGNORED;
	if (event_shift_is_down (event)) c = c -  32;
	/* this will give a non-{lrt} for unshifted keys */
	if (event_ctrl_is_down  (event)) c = c -  64;
	if (event_meta_is_down  (event)) c = c + 128;
#ifdef WANT_CAPS_LOCK
/* set a toggle and relabel window so T1 can act like caps-lock */
	if (event_id(event) == KEY_TOP(1))
	  {
	    /* make a frame label with and without CAPS */
	    strcpy (buffer, Caps);
	    title = &buffer[CAPS_LEN];
	    strncpy (title, (char *)window_get (frame, FRAME_LABEL),
		     BUFFER_SIZE - CAPS_LEN);
	    buffer[BUFFER_SIZE] = (char) 0;
	    if (strncmp (title, Caps, CAPS_LEN) == 0)
	      title += CAPS_LEN; 		 /* already Caps */
	    caps_lock =  (caps_lock ? 0 : CAPS_LEN);
	    window_set(frame, FRAME_LABEL, (title -= caps_lock), 0);
	    return NOTIFY_IGNORED;
	  }
#endif
	ttysw_input (tty_win, key_prefix, k_prefix_length);
	sprintf (buffer, "%c%c", d, c);
	ttysw_input(tty_win, buffer, strlen(buffer));

	return NOTIFY_IGNORED;
      }
  }
  if ((event_is_ascii(event) || event_is_meta(event))
      && event_is_up(event)) return NOTIFY_IGNORED;
#ifdef WANT_CAPS_LOCK
/* shift alpha chars to upper case if toggle is set */
  if ((caps_lock) && event_is_ascii(event)
      && (event_id(event) >= 'a') && (event_id(event) <= 'z'))
    event_set_id(event, (event_id(event) - 32));
/* crufty, but it works for now. is there an UPCASE(event)? */
#endif
#ifndef NO_META_BIT
/* under Openwindows/X, the meta bit is not set in the key event,
 * emacs expects this so we add it in here:
 */
  if (event_is_ascii(event) && event_meta_is_down(event))
    event_set_id(event, 128 | event_id(event));
#endif
  return notify_next_event_func (window, event, arg, type);
}

main (argc, argv)
     int argc;
     char **argv;
{
  int error_code;	/* Error codes */

#ifdef JLE
  setlocale(LC_ALL, "");
#endif JLE

  if(getenv("DEBUGEMACSTOOL"))
    console = fdopen (console_fd = open("/dev/console",O_WRONLY), "w");

  putenv("IN_EMACSTOOL=t");	/* notify subprocess that it is in emacstool */

  if (putenv("TERM=sun") != 0)	/* TTY_WIN will be a TERM=sun window */
    {fprintf (stderr, "%s: Could not set TERM=sun, using `%s'\n",
	     argv[0], (char *)getenv("TERM")) ;};
  /*
   * If TERMCAP starts with a slash, it is the pathname of the
   * termcap file, not an entry extracted from it, so KEEP it!
   * Otherwise, it may not relate to the new TERM, so Nuke-It.
   * If there is no TERMCAP environment variable, don't make one.
   */
  {
    char *termcap ;	/* Current TERMCAP value */
    termcap = (char *)getenv("TERMCAP") ;
    if (termcap && (*termcap != '/'))
      {
	if (putenv("TERMCAP=") != 0)
	  {fprintf (stderr, "%s: Could not clear TERMCAP\n", argv[0]) ;} ;
      } ;
  } ;

  /* find command to run as subprocess in window */
  if (!(argv[0] = (char *)getenv("EMACSTOOL")))	/*  Set emacs command name */
      argv[0] = emacs_name;
  /* Emacstool recognizes two special args: -rc <file> and -bold <bold-name> */
  for (argc = 1; argv[argc]; argc++)		/* Use last one on line */
    {
      if(!(strcmp ("-rc", argv[argc])))		/* Override if -rc given */
	{int i = argc;
	 argv[argc--]=0;		/* kill the -rc argument */
	 if (argv[i+1]) {	/* move to argv[0] and squeeze the rest */
	   argv[0]=argv[i+1];
	   for (; argv[i+2]; (argv[i]=argv[i+2],argv[++i]=0));
	 }
       }

      if (!(strcmp ("-bold", argv[argc])))
	  {int i = argc;
	   argv[argc--]=0;		/* kill the -bold argument */
	   if (argv[i+1]) {	/* move to bold_name and squeeze the rest */
	       bold_name = argv[i+1];
	       for (; argv[i+2]; (argv[i]=argv[i+2],argv[++i]=0));
	   }
       }
  };

  strcpy (buffer, title);
  strncat (buffer, argv[0],		 /* append run command name */
	   (BUFFER_SIZE - (strlen (buffer)) - (strlen (argv[0]))) - 1);

  error_code = interpose_on_window(argc,argv);
  if (error_code != 0) {		/* Barf */
      fprintf (stderr, "notify_interpose_event_func returns %d.\n", error_code);
      exit (1);
  }

#ifdef XVIEW
  xv_main_loop (frame);                  /* And away we go */
#else
  window_main_loop (frame);
#endif XVIEW
}

#ifdef XVIEW
int interpose_on_window(argc,argv)
    int argc;
    char **argv;
{
#ifndef TTERM
    int i, font_width_adjust = 1; /* hackery, and heuristics */
    /* if -Wt is not supplied, then font comes out as lucida-14 (width=8)
     * rather than the screen.r.12 (width=7) typically used
     * this hack attempts to workaround it.
     * could use a env var EMACSTOOL_DEFAULT_FONT_WIDTH instead */
    for (i = 1; argv[i]; i++) {
	if (!(strcmp ("-Wt", argv[i])))
	    {font_width_adjust = 0;
	     if (console_fd) fprintf(console, "-Wt = %d\n", font_width_adjust);
	     break;}
    }
#endif TTERM
    /* initialize Xview, and strip window args */
    xv_init(XV_INIT_ARGC_PTR_ARGV, &argc, argv, 0);

    /* do this first, so arglist can override it */
    frame_icon = icon_create (ICON_LABEL, "Emacstool",
			      ICON_IMAGE, &icon_image,
			      0);

    /* Build a frame to run in */
    frame = xv_create ((Xv_Window)NULL, FRAME,
		       FRAME_LABEL, buffer,
		       FRAME_ICON, frame_icon,
		       0);

    /* Create a tty with emacs in it */
    tty_win = xv_create (frame, SWTYPE, WIN_IS_CLIENT_PANE,
			 TTY_QUIT_ON_CHILD_DEATH, TRUE,
			 TTY_BOLDSTYLE, TTYSW_BOLD_INVERT,
			 TTY_ARGV, argv,
			 0);

    if (bold_name) {
	(void)xv_set(tty_win, TTY_BOLDSTYLE_NAME, bold_name, 0);
    }

    {
	Xv_font font;		/* declare temp font variable */
	font = (Xv_font)xv_get (tty_win, XV_FONT);
	font_height = (int)xv_get (font, FONT_DEFAULT_CHAR_HEIGHT);
	font_width  = (int)xv_get (font, FONT_DEFAULT_CHAR_WIDTH);
    }
    if (console_fd) fprintf(console, "Width = %d\n", font_width);

#ifndef TTERM
    font_width -= font_width_adjust; /* A guess! font bug in ttysw*/
#else
    /* make the termsw act as a tty */
    xv_set(tty_win, TERMSW_MODE, TTYSW_MODE_TYPE, 0);
    /* termsw has variable offset depending on scrollbar size/location */
    left_margin = (int)xv_get (tty_win, TEXTSW_LEFT_MARGIN);
#endif TTERM

    tty_view = (Xv_Window) xv_get (tty_win, OPENWIN_NTH_VIEW, 0);
    xv_set(tty_view,
	   WIN_CONSUME_EVENTS,
	   WIN_MOUSE_BUTTONS, WIN_UP_EVENTS,
	   ACTION_ADJUST, ACTION_MENU,
	   WIN_ASCII_EVENTS,
	   WIN_LEFT_KEYS, WIN_TOP_KEYS, WIN_RIGHT_KEYS,
	   0,
	   0);
    /* Interpose my event function */
    return (int) notify_interpose_event_func
	(tty_view, input_event_filter_function, NOTIFY_SAFE);
}
#else
int interpose_on_window (argc, argv)
 int argc;
 char **argv;
{
    /* do this first, so arglist can override it */
    frame_icon = icon_create (ICON_LABEL, "Emacstool",
			      ICON_IMAGE, &icon_image,
			      0);

    /* Build a frame to run in */
    frame = window_create ((Window)NULL, FRAME,
			   FRAME_LABEL, buffer,
			   FRAME_ICON, frame_icon,
			   FRAME_ARGC_PTR_ARGV, &argc, argv,
			   0);

    /* Create a tty with emacs in it */
    tty_win = window_create (frame, TTY,
			     TTY_QUIT_ON_CHILD_DEATH, TRUE,
			     TTY_BOLDSTYLE, TTYSW_BOLD_INVERT,
			     TTY_ARGV, argv,
			     0);

    if (bold_name) {
	(void)window_set(tty_win, TTY_BOLDSTYLE_NAME, bold_name, 0);
    }

    /* ttysw uses pf_default, one must set WIN_FONT explicitly */
                       window_set (tty_win, WIN_FONT, pf_default(), 0);
    font_height = (int)window_get (tty_win, WIN_ROW_HEIGHT);
    font_width  = (int)window_get (tty_win, WIN_COLUMN_WIDTH);

    tty_view = tty_win;
    window_set(tty_view,
	       WIN_CONSUME_PICK_EVENTS,
	       WIN_STOP,
	       WIN_MOUSE_BUTTONS, WIN_UP_EVENTS,
	       /* LOC_WINENTER, LOC_WINEXIT, LOC_MOVE, */
	       0,
	       WIN_CONSUME_KBD_EVENTS,
	       WIN_STOP,
	       WIN_ASCII_EVENTS,
	       WIN_LEFT_KEYS, WIN_TOP_KEYS, WIN_RIGHT_KEYS,
	       /* WIN_UP_ASCII_EVENTS, */
	       0,
	       0);
    /* Interpose my event function */
    return (int) notify_interpose_event_func
	(tty_view, input_event_filter_function, NOTIFY_SAFE);
}
#endif XVIEW

/* arch-tag: 7a2e7105-c059-418a-b3d9-5b5de96abb4e
   (do not change this comment) */
